kvserver: gate writing the new sticky GCHint

pav-kv opened the pull request:
This PR makes the new behaviour of `GCHint` introduced in #110078 conditional on a default-off cluster setting in 23.1, and enabled in 23.2 by a version gate.

Since it is late to enable this behaviour in 23.1 (risk of backwards incompatibility), hide it behind a default-off cluster setting. In 23.2, it will be enabled by default, and the cluster setting will be deprecated.

The new `GCHint` behaviour is likely backwards compatible, but we are hiding it behind a setting for extra safety.

The safest moment to enable this cluster setting is when there is some confidence that the cluster binaries will not rollback to previous patch versions of 23.1. The risk exists only in mixed-version state in which some 23.1 binaries don't know the new `GCHint` fields, and some do.

Epic: none

Release note (ops change): introduce a default-off cluster setting `kv.gc.sticky_hint.enabled` which helps expediting garbage collection after range deletions, such as when a SQL table or index is dropped.

Release note (general change): set the default behaviour for `kv.gc.sticky_hint.enabled` cluster setting to enabled since 23.2. The setting is deprecated in 23.2 going forward.

pav-kv committed:
stateloader: remove redundant bool return value

StateLoader.SetGCHint returns true iff err == nil. We can safely remove
the bool return value.

Epic: none
Release note: none

pav-kv committed:
kvserver: write new GCHint fields under cluster setting

Since it is late to enable this behaviour in 23.1 (risk of backwards
incompatibility), hide it behind a default-off cluster setting. In 23.2,
it will be enabled by default, and the cluster setting will be
deprecated.

The new GCHint behaviour is likely backwards compatible, but we are
hiding it behind a setting for extra safety.

The safest moment to enable this cluster setting is when there is some
confidence that the cluster binaries will not rollback to previous patch
versions of 23.1. The risk exists only in mixed-version state in which
some 23.1 binaries don't know the new GCHint fields, and some do.

Epic: none
Release note (ops change): introduce a default-off cluster setting
`kv.gc.sticky_hint.enabled` which helps expediting garbage collection
after range deletions, such as when a SQL table or index is dropped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112948)
<!-- Reviewable:end -->

pav-kv committed:
batcheval: write new GCHint fields under version gate

Epic: none
Release note (general change): set the default behaviour for
`kv.gc.sticky_hint.enabled` cluster setting to enabled since 23.2. The
setting is deprecated in 23.2 going forward.

pavelkalinnikov commented:
@erikgrinaker Ready for a first pass. LMK if more commenting/documenting is needed. Also, I think we should add some minimal testing of the on/off behaviour of the cluster setting.

pavelkalinnikov commented:
All commits of this PR (except the 23.2 version gate) are intended to be added to the backport #110643.

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_delete_range.go:
```diff
@@ -14,19 +14,29 @@ import (
+var enableStickyGCHint = settings.RegisterBoolSetting(
+	settings.SystemOnly,
+	"kv.gc.sticky_hint.enabled",
+	"enable writing sticky GC hints which expedite garbage collection after range deletions",
```
```suggestion
	"enable writing sticky GC hints which expedite garbage collection after schema changes",
```

pavelkalinnikov replied:
Done.

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_end_transaction.go:
I think we should respect the new hint if it has already been written, regardless of the setting and gate. It's a lot easier to reason about, both for us and users, and can also be racy since version gates aren't enabled atomically across cluster nodes. Otherwise, existing hints will sometimes be respected but sometimes not, depending on whether there's been a range merge in the meanwhile.

pavelkalinnikov replied:
Sounds reasonable. Will modify the code as follows (if I understand the proposal right):
- Check the `cluster setting || version gate` condition only in `cmd_delete_range.go`, where the `GCHint` originates from.
- Here in `cmd_end_transaction.go`, only look at the new "GC timestamp" fields (to determine whether this is a "new" hint), and branch accordingly.

This has the "atomicity" benefit, which relies on the fact that if a GCHint has been written in the new format, then the node on which it "originated" had the gate enabled, which means that the entire cluster already uses the new binary which "understands" the new hint.

pavelkalinnikov replied:
> - Here in cmd_end_transaction.go, only look at the new "GC timestamp" fields (to determine whether this is a "new" hint), and branch accordingly.

Actually, sounds like we don't need to branch here. We just use the `Merge` function. It will automatically "respect" the new hints, but will also work for old hints.

erikgrinaker replied:
Yep, sounds good.

pavelkalinnikov replied:
Done.

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_delete_range.go:
```diff
@@ -14,19 +14,29 @@ import (
+var enableStickyGCHint = settings.RegisterBoolSetting(
+	settings.SystemOnly,
+	"kv.gc.sticky_hint.enabled",
+	"enable writing sticky GC hints which expedite garbage collection after schema changes",
```
Should probably say "(ignored in 23.2)".

pavelkalinnikov replied:
Added `(ignored and assumed true in 23.2)`.

erikgrinaker reviewed (approved):
Let's drop the first commit in the 23.1 backport, to minimize the change.

pavelkalinnikov commented:
bors r=erikgrinaker
